torch/csrc/jit/register_prim_ops.cpp

```diff
@@ -1009,14 +1019,232 @@ Operator(                                                                      \
+  if (input_dim == 5 && mode == "area")
+    return at::adaptive_avg_pool3d(input, _output_size(input, 3, size, scale_factors));
+  if (input_dim == 3 && mode == "linear")
+    return at::upsample_linear1d(input, _output_size(input, 1, size, scale_factors), *align_corners);
+  if (input_dim == 3 && mode == "bilinear")
+    throw std::runtime_error("Got 3D input, but bilinear mode needs 4D input");
+  if (input_dim == 3 && mode == "trilinear")
+    throw std::runtime_error("Got 3D input, but trilinear mode needs 5D input");
+  if (input_dim == 4 && mode == "linear")
+    throw std::runtime_error("Got 4D input, but linear mode needs 3D input");
+  if (input_dim == 4 && mode == "bilinear")
+    return at::upsample_bilinear2d(input, _output_size(input, 2, size, scale_factors), *align_corners);
+  if (input_dim == 4 && mode == "trilinear")
+    throw std::runtime_error("Got 4D input, but trilinear mode needs 5D input");
+  if (input_dim == 5 && mode == "linear")
+    throw std::runtime_error("Got 5D input, but linear mode needs 3D input");
+  if (input_dim == 5 && mode == "bilinear")
+    throw std::runtime_error("Got 5D input, but bilinear mode needs 4D input");
+  if (input_dim == 5 && mode == "trilinear")
+    return at::upsample_trilinear3d(input, _output_size(input, 3, size, scale_factors), *align_corners);
+
+  AT_ERROR("Input Error: Only 3D, 4D and 5D input Tensors supported",
+    " (got ", input_dim, "D) for the modes: nearest | linear | bilinear | trilinear",
+    " (got ", mode, ") ");
+}
+
+Operation interpolate_op(const Node* n) {
+  return [](Stack& stack) {
+    at::Tensor input;
+    IValue size;
+    IValue scale_factors;
+    std::string mode;
+    IValue align_corners;
+    pop(stack, input, size, scale_factors, mode, align_corners);
+    at::Tensor res = interpolate(input, size, scale_factors, mode, align_corners.toOptional<bool>());
+    push(stack, res);
+    return 0;
+  };
+}
+
+// interpolate takes in float & float[] for scale factor
+// upsample takes in int & int[], so convert the ints to floats before
+// passing on to the interpolate op
+IValue convert_int_or_int_list_to_double(IValue int_ivalue) {
+  IValue scale_factor_double;
+  if (int_ivalue.isInt()) {
+    scale_factor_double = static_cast<double>(int_ivalue.toInt());
+  } else if (int_ivalue.isIntList()) {
+    auto int_list = int_ivalue.toIntListRef();
+    std::vector<double> double_vec(int_list.begin(), int_list.end());
+    scale_factor_double = double_vec;
+  }
+  return scale_factor_double;
+}
+
+Operation upsample_nearest_op(const Node* n) {
+  return [](Stack& stack) {
+    at::Tensor input;
+    IValue size;
+    IValue scale_factor_int;
+    pop(stack, input, size, scale_factor_int);
+    IValue scale_factor_double = convert_int_or_int_list_to_double(scale_factor_int);
+    at::Tensor res = interpolate(input, size, scale_factor_double, "nearest", c10::nullopt);
+    push(stack, res);
+    return 0;
+  };
+}
+
+Operation upsample_op(const Node* n) {
```

Comment:
Shouldn't upsample/upsample_bilinear calls interpolate? Can we just make interpolate as a builtin, and fix the upsample/upsample_bilinear in functional instead?

Reply:
They do call interpolate, scale factor is a double for interpolate and an int for upsample. So the parameters need to be converted before the call to interpolate.

Reply:
You can just do float(scale) for upsample functions in functional.py?

Reply:
The parameters are overloaded, it accepts both ?int, and ?intlist, and there isn't a fixed length the int can be broadcasted to. I can look into implementing it in register_builtins, I think I had an issue doing that but will try again.

Reply:
This looks good in the case that we already have interpolate registered here. I think a better solution might be: in upsample functions in the nn/functional.py, check if the scale_factor is a list, and transform the scale_factor to float using `float(scale_factor)` or `[float(i) for i in scale_factor]`, then pass into the interpolate function. This way we can keep a single interpolate function in c++, but this requires `isinstance` which we don't currently have in JIT yet... and the cast might not be good enough still.